// datamcfly/index.js
var urlParser = require('url');
var http = require('http');
var md5 = require('MD5');

/*
	DataMcFly class
	@database {String}
    @collection {String}
	@key {String} :: get your key from http://app.datamcfly.com/
*/
function DataMcFly(database, collection, apikey) {

	this.apikey = apikey;
	this.query = {};
	this.database = database;
	this.collection = collection;
	this.currentItem;
	this.room = md5( database + '/' + collection ); 		//	this will be a hash of the room..
	this.uri = urlParser.parse('http://api.datamcfly.com/apps/');
	this.push_uri = urlParser.parse('http://push.datamcfly.com');
	this.startWebSocket( this.room );
};

// ======================================================
// FUNCTIONS
// ======================================================

/*
	Check if string is JSON object
	@value {String}
	return {Object}
*/
function parseJSON(value) {
	if (isJSON(value))
		return JSON.parse(value);
	return {};
};

function isJSON(value) {
	if (value.length === 1)
		return false;
	var a = value[0];
	var b = value[value.length - 1];
	return (a === '"' && b === '"') || (a === '[' && b === ']') || (a === '{' && b === '}');
};

/*
	Object to URL params
	@obj {Object}
	return {String}
*/
function toParams(obj, apikey) {

	if (typeof(obj) === 'undefined' || obj === null)
		return '?apiKey=' + apikey;

	var buffer = [];

	var arr = Object.keys(obj);

	arr.forEach(function(o) {

		var value = obj[o];
		var name = o.toLowerCase().replace(/-/g, '');

		switch (name) {
			case 'q':
			case 'query':
			case 'where':
				buffer.push('q=' + encodeURIComponent(JSON.stringify(value)));
				break;		

			case 'include':

				if (typeof(value) === 'array') {
					var prepare = {};
					value.forEach(function(o) {
						prepare[o] = 1;
					});
					value = prepare;
				}

				buffer.push('f=' + encodeURIComponent(JSON.stringify(value)));
				break;

			case 'exclude':

				if (typeof(value) === 'array') {
					var prepare = {};
					value.forEach(function(o) {
						prepare[o] = 0;
					});
					value = prepare;
				}

				buffer.push('f=' + encodeURIComponent(JSON.stringify(value)));
				break;

			case 'f':
				buffer.push('f=' + encodeURIComponent(JSON.stringify(value)));
				break;

			case 'l':
			case 'max':
			case 'top':
			case 'limit':
			case 'take':
				buffer.push('l=' + value);
				break;

			case 'skip':
			case 'sk':
				buffer.push('sk=' + value);
				break;

			case 'asc':

				if (typeof(value) === 'array') {
					var prepare = {};
					value.forEach(function(o) {
						prepare[o] = 1;
					});
					value = prepare;
				}

				buffer.push('s=' + encodeURIComponent(JSON.stringify(value)));
				break;

			case 'desc':

				if (typeof(value) === 'array') {
					var prepare = {};
					value.forEach(function(o) {
						prepare[o] = -1;
					});
					value = prepare;
				}

				buffer.push('s=' + encodeURIComponent(JSON.stringify(value)));
				break;

			case 's':
			case 'sort':
			case 'sorting':
				buffer.push('s=' + encodeURIComponent(JSON.stringify(value)));
				break;

			case 'c':
			case 'count':
				buffer.push('c=' + value.toString().toLowerCase());
				break;

			case 'fo':
			case 'first':
				buffer.push('fo=' + value.toString().toLowerCase());
				break;

			case 'm':
			case 'updateall':
			case 'all':
				buffer.push('m=' + value.toString().toLowerCase());
				break;

			case 'u':
			case 'upsert':
				buffer.push('u=' + value.toString().toLowerCase());
				break;

			default:
				break;
		};
	});

	var params = buffer.join('&');
//	console.log( '?' + (params.length > 0 ? params + '&' : '') + 'apiKey=' + apikey );
	return '?' + (params.length > 0 ? params + '&' : '') + 'apiKey=' + apikey;
};

// ======================================================
// PROTOTYPES
// ======================================================

/*
	Internal function
	@path {String}
	@method {String}
	@data {String or Object or Array}
	@params {Object}
	@callback {Function} :: function(error, object)
	return {self}
*/

DataMcFly.prototype.startWebSocket = function ( channel ){
	this.socket = require('socket.io-client')('http://push.datamcfly.com');

	var data = {};
	data.room = this.room;
	data.apiKey = this.apikey;
	data.db = this.database;
	data.collection = this.collection;

/*
	this.socket.on('connect', function(){});
	this.socket.on('event', function(data){});
	this.socket.on('disconnect', function(){});
*/

//	console.log( "subscribing to channel " + channel );
	this.socket.emit('subscribe', data );

	this.socket.on("connect", function() {
		console.log( "Connected" );
	}).on("disconnect", function() {
		console.log( "Disconnected" );
	}).on("connecting",function(){
		console.log( "Connecting" );
	});
  
	this.socket.on('connected', function (data) {
		console.log( data );
	});
	this.socket.on('status', function (data) {
		console.log( data );
	});

//	this.socket.emit('status', "Hello Status");
	
	return true;
};

//	Set notifications when event is returned...
/*
DataMcFly.prototype.on = function( event, callback ){
	if( event == 'value' ){
		return this.documents({},callback);
	}
	this.socket.on( event, callback );
};
*/
DataMcFly.prototype.where = function( where ){
	this.query.q = where;
	return this;
}

DataMcFly.prototype.fields = function( field ){
	//	return only the field we passed as $field...
	this.query.f = field;
	return this;
}

DataMcFly.prototype.skip = function( n ){
	this.query.sk = n;
	return this;
}

DataMcFly.prototype.orderBy = function( o ){
	this.query.s = o;
	return this;
}

DataMcFly.prototype.limit = function( n ){
	this.query.l = n;
	return this;
}
DataMcFly.prototype.limitToFirst = function( n ){
	this.query.l = n;
//	this.query.s = {"_id": 1};
	return this;
}
DataMcFly.prototype.limitToLast = function( n ){
	this.query.l = n;
	this.query.s = {"_id": -1};
	return this;
}

DataMcFly.prototype.onDisconnect = function( callback ){
//	window.onbeforeunload = callback;
/*
	window.addEventListener("beforeunload", function (e) {
		callback();
		return null;
	});
*/
	var myEvent = window.attachEvent || window.addEventListener;
	var chkevent = window.attachEvent ? 'onbeforeunload' : 'beforeunload'; /// make IE7, IE8 compitable
	myEvent(chkevent, function(e) { // For >=IE7, Chrome, Firefox
		callback();
	});
}

DataMcFly.prototype.on = function( key, callback ){
	if( key == 'value' ){
		this.documents(this.query,callback);
		//	once the initial query finishes, listen for any new records...
//		this.on('added', callback);
	}else{
		var self = this;

		this.socket.on( key, function(res){
			if( key == 'added' || key == 'changed' || key == 'removed' || key == 'online' ){
				var data = self.processData( res );
				self.currentItem = data;
			}else{
				var data = res;
			}
			callback( data );
		});
	}
};

DataMcFly.prototype.once = function( key, callback ){
	if( key == 'value' ){
		//	check query based on other functions.. 
		this.listDocuments(callback,this.query);
	}else{
		var self = this;
		
		this.socket.on( key, function(res){
			if( key == 'added' || key == 'changed' || key == 'removed' || key == 'online' ){
				var data = self.processData( res );
				self.currentItem = data;
			}else{
				var data = res;
			}
			callback( data );
			return true;
		});
	}
};

//	Send message to notification server...
DataMcFly.prototype.trigger = function(event, message){
	http.get(this.push_uri+'/emit/' + this.room + '/' + event + '/' + message, function(res) {
//		console.log("Got response: " + res.statusCode);
	}).on('error', function(e) {
//		console.log("Got error: " + e.message);
	});
};
DataMcFly.prototype.emit  = function(event, message) {
	http.get(this.push_uri+'/emit/' + this.room + '/' + event + '/' + message, function(res) {
//		console.log("Got response: " + res.statusCode);
	}).on('error', function(e) {
//		console.log("Got error: " + e.message);
	});
};


DataMcFly.prototype.connect = function(path, method, data, params, callback) {

	var self = this;

	if (path[0] === '/')
		path = path.substring(1);

	var uri = self.uri;
	var type = typeof(data);
	var isObject = type === 'object' || type === 'array';

	var headers = {};

	headers['Content-Type'] = isObject ? 'application/json' : 'text/plain';
	headers['X-DataMcFly-API-Key'] = self.apikey;
	var location = '';

	if (path[0] === '#')
		location = path.substring(1);
	else
		location = uri.pathname + path;

	var options = { 
			protocol: uri.protocol, 
			auth: uri.auth, 
			method: method || 'GET', 
			hostname: uri.hostname, 
			port: uri.port, 
			path: location + toParams(params, self.apikey), 
			agent:false, 
			headers: headers 
	};

//	console.log( location + toParams(params, self.apikey));

	var response = function (res) {
		var buffer = '';

		res.on('data', function(chunk) {
			buffer += chunk.toString('utf8');
		})

		req.setTimeout(exports.timeout, function() {
			callback(null,new Error('timeout'));
		});

		res.on('end', function() {
			var data = parseJSON(buffer.trim());
			var error = null;

			if (res.statusCode > 200) {				
				error = new Error(res.statusCode + ' (' + (data.message || '') + ') ');
				data = null;
			}

			var data = self.processData( data );

			self.currentItem = data;
			callback(data,error);
		});
	};

	var con = options.protocol === 'http:' ? http : http;
	var req = callback ? con.request(options, response) : con.request(options);

	req.on('error', function(err) {
		callback(null,err);
	});

	if (isObject)
		req.end(JSON.stringify(data));
	else
		req.end();

	return self;
};

/*
	DataMcFly command
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.collections = function(cb) {
	var self = this;
	return self.connect(self.database + '/collections', 'GET', null, null, cb);
};

/*
	DataMcFly command
	@params {Object}
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.get = function(cb,params) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'GET', null, params, cb);
};
DataMcFly.prototype.listDocuments = function(cb, params) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'GET', null, params, cb);
};
DataMcFly.prototype.documents = function(params, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'GET', null, params, cb);
};

/*
	DataMcFly command
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.databases = function(cb) {
	return this.connect('', 'GET', null, null, cb);
};

/*
	DataMcFly command
	@documents {Object or Object array}
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.set = function(documents, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'POST', documents, null, cb);
};

DataMcFly.prototype.push = function(documents, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'POST', documents, null, cb);
};

DataMcFly.prototype.insert = function(documents, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'POST', documents, null, cb);
};

/*
	DataMcFly command
	@condition {Object or Object array}
	@params {Object}
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.update = function(condition, params, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection, 'PUT', condition, params, cb);
};

/*
	DataMcFly command
	@id {String or Number}
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.findId = function(id, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection + '/' + id, 'GET', null, null, cb);
};

/*
	DataMcFly command
	@id {String or Number}
	@document {Object}
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.updateId = function(id, document, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection + '/' + id, 'PUT', document, null, cb);
};

/*
	DataMcFly command
	@collection {String}
	@id {String or Number}
	@cb {Function} :: function(error, object)
	return {DataMcFly}
*/
DataMcFly.prototype.remove = function(id, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection + '/' + id, 'DELETE', null, null, cb);
};
DataMcFly.prototype.deleteDocument = function(id, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection + '/' + id, 'DELETE', null, null, cb);
};
DataMcFly.prototype.deleteId = function(id, cb) {
	var self = this;
	return self.connect(self.database + '/collections/' + this.collection + '/' + id, 'DELETE', null, null, cb);
};

DataMcFly.prototype.transaction = function( updateFunction, cb){
	callback = cb || function(){};
	var c = {
		update: updateFunction,
		status: null
	};
	var e = this.currentItem.value();
	var e = e[0];
	var d = c.update( e );
	if ( is_void(d) ) {
		console.log("transaction failed: Data returned " + d);
		c.status = 1;
	}else{
		if( is_object( d ) ){
//	merge the arrays together...
			var k = merge( e, d );
			this.push( k, cb );
		}else{
//	returned null.. so delete...
			var id = e._id;
//			delete the record since it was NULL..
			this.deleteId( e._id, cb );
		}
		console.log( d );
	}
};

//	format return values into a object with helper functions...
DataMcFly.prototype.processData = function ( data ){
	var self = this;
	var retVal = [];
	var raw = data;

	var IS_JSON = true;
	try{
		var json = JSON.parse( data );
	}catch(err){
		IS_JSON = false;
	}

	if( IS_JSON ){
		var toProcess = json;
	}else{
		var toProcess = data;
	}

	if( !toProcess.length ){
		//	single variable...
		var obj = {
			'data': toProcess,
			value : function() {
				return this.data;
			},
			key : function(){
				return this.data._id;
			},
			ref : function(){
				return self;
			}
		};
		var Processed = obj;
	}else{
		for (var i in toProcess) {
			var value = toProcess[i];
			var obj = {
				'data': value,
				value : function() {
					return this.data;
				},
				key : function(){
					return this.data._id;
				},
				ref : function(){
					return self;
				}
			};
			retVal[i] = obj;
		}
		var Processed = retVal;
	}				
	
//	single variable...
	var data = {
		'data': Processed,
		'raw': raw,
		value : function() {
			if( this.count() > 1 ){
				return this.data;
			}else{
				return this.data.value();
			}
		},
		key : function(){
			if( this.count() > 1 ){
				return null;
			}else{
				return this.data.key();
			}
		},
		first : function(){
			if( this.count() > 0 ){
				var d = this.data[ 0 ];
				return d;
			}else{
				return this.data;				
			}
		},
		count : function(){
			if (typeof this.data.length == 'undefined' ){
				return null;
			}else{
				return this.data.length;	
			}
		},
		forEach : function( cb ){
			// iterate through each record returned...
			for (var i in this.data ) {
				var d = this.data[i];

				cb( d );
			}
		},
		ref : function(){
			return self;
		}
	};
	
	return data;
};

//	Utility Functions ---------------------------------------------------------------

var hb = function() {
	var a = 1;
	return function() {
		return a++
	}
}();

function is_object(a) {
	var b = typeof a;
	return "object" == b && null != a || "function" == b
}

function is_void(a) {
//	return (typeof a === 'undefined');
	return void 0 === a;
}

function merge() {
	var obj = {}, i = 0, il = arguments.length, key;
	for (; i < il; i++) {
		for (key in arguments[i]) {
			if (arguments[i].hasOwnProperty(key)) {
				obj[key] = arguments[i][key];
			}
		}
	}
	return obj;
};



// ======================================================
// EXPORTS
// ======================================================

exports.timeout = 10000;
exports.DataMcFly = exports.datamcfly = exports.mongodb = DataMcFly;

/*
	DataMcFly class
	@connectionString {String} :: url address
*/
exports.init = function(db, collection, apikey) {
	return new DataMcFly(db, collection, apikey);
};
